### PR TITLE
WIP : Feature preprocessing

### PR DESCRIPTION
**Pull request by vitornesello**
Todo:
 - [x] style : tabs -> spaces
 - [x] fix termination status bug : #136 
 - [x] fix sp vars bound changes : #121 
 - [x] style : Indentation
 - [x] fix remaining bugs

### CODE REVIEW

**Comment by vitornesello** [comment]
Now we just need to improve the style of preprocessing.jl. Then we are good to merge in master 💪

File: src/algorithms/preprocessing.jl [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+                                         alg_data.cur_min_slack[constr.id], -coef*cur_lb, Inf))
+    end
+end
+
+function analyze_constraint(alg_data::PreprocessData, constr::Constraint, form::Formulation)
+    if constr.duty <:AbstractMasterConstr
+       var_filter = _rep_of_orig_var_ 
+    else
+       var_filter = (var -> (var.duty == DwSpPricingVar))
+    end
```
Comment (guimarqu) [inline src/algorithms/preprocessing.jl]: maybe it's better to create a function that returns the good filtering function using multiple dispatch

File: src/algorithms/preprocessing.jl [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+     return false
+ end
+
+function propagation(alg_data::PreprocessData)
+     while !isempty(alg_data.stack)
+         (constr, form) = pop!(alg_data.stack)
+         alg_data.constr_in_stack[constr.id] = false
+
+         if alg_data.printing
+             println("constr $(getname(constr)) $(typeof(constr)) popped")
```
Comment (guimarqu) [inline src/algorithms/preprocessing.jl]: You should avoid string interpolation

File: src/algorithms/preprocessing.jl [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+        return (is_ub, compute_new_bound(alg_data.nb_inf_sources_for_max_slack[constr.id],
+                                         alg_data.cur_max_slack[constr.id], -coef*cur_ub, -Inf))
+    else
+        is_ub = true
+        return (is_ub, compute_new_bound(alg_data.nb_inf_sources_for_min_slack[constr.id], 
+                                         alg_data.cur_min_slack[constr.id], -coef*cur_lb, Inf))
+    end
+end
+
+function analyze_constraint(alg_data::PreprocessData, constr::Constraint, form::Formulation)
```
Comment (guimarqu) [inline src/algorithms/preprocessing.jl]: I don't like the name because it does not give a clue on what the method does

File: src/node.jl [hunk 4]
```diff
@@ -157,50 +175,74 @@ function apply_branch!(f::Reformulation, b::Branch)
...
-            setlb!(owner_form, getvar(owner_form, id), getlb(data))
-            setub!(owner_form, getvar(owner_form, id), getub(data))
-            @logmsg LogLevel(-3) string("New lower bound is ", getcurlb(var))
-            @logmsg LogLevel(-3) string("New upper bound is ", getcurub(var))
+    return
+end
+
+function apply_data!(form::Formulation, constr::Constraint, constr_data::ConstrData)
+    # Rhs
+    if (getcurrhs(constr) != getrhs(constr_data))
```
Comment (guimarqu) [inline src/node.jl]: useless parenthesis

File: src/node.jl [hunk 5]
```diff
@@ -157,50 +175,74 @@ function apply_branch!(f::Reformulation, b::Branch)
...
+
+function reset_to_record_state_of_father!(form::Formulation, record::NodeRecord)
+    @logmsg LogLevel(-2) "Checking variables"
+    reset_var_constr(form, record.active_vars, getvars(form))
+    @logmsg LogLevel(-2) "Checking constraints"
+    reset_var_constr(form, record.active_constrs, getconstrs(form))
+    return
+end
+
+function reset_var_constr(form::Formulation, active_var_constrs, var_constrs_in_formulation)
```
Comment (guimarqu) [inline src/node.jl]: missing ! because the method looks like to modify the formulation

File: src/optimizationresults.jl [hunk 6]
```diff
@@ -30,3 +52,37 @@ nbprimalsols(res::OptimizationResult) = length(res.primal_sols)
 nbdualsols(res::OptimizationResult) = length(res.dual_sols)
 getbestprimalsol(res::OptimizationResult) = res.primal_sols[1]
 getbestdualsol(res::OptimizationResult) = res.dual_sols[1]
+setprimalbound!(res::OptimizationResult, b::PrimalBound) = res.primal_bound = b
+setdualbound!(res::OptimizationResult, b::DualBound) = res.dual_bound = b
+setterminationstatus(res::OptimizationResult, status::TerminationStatus) = res.termination_status = status
```
Comment (guimarqu) [inline src/optimizationresults.jl]: !

File: src/algorithms/preprocessing.jl (hunk 3 above)
Comment (tbulhoes) [reply src/algorithms/preprocessing.jl]: This functions loops over all variables in the constraint trying to change their bounds. If a bound changes, new constraints are added to the stack. It is called by propagation. I am not sure whether it is correct to call it "propagate_constraint". maybe @FD-V has a better name for it.

File: src/algorithms/preprocessing.jl [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+        if (update_min_slack(alg_data, constr, form, false, 0.0) 
+            || update_max_slack(alg_data, constr, form, false, 0.0))
+            return true
+        end
+    end
+
+    return false
+ end
+
+function initialize_constraint(alg_data::PreprocessData, constr::Constraint, form::Formulation)
```
[the remark this answers is not in the document]
Comment (tbulhoes) [reply src/algorithms/preprocessing.jl]: I have added ! to all functions with side effects (almost all of them).

File: src/algorithms/preprocessing.jl [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+        alg_data.nb_inf_sources_for_max_slack[constr.id] -= 1
+    end
+
+    nb_inf_sources = alg_data.nb_inf_sources_for_max_slack[constr.id]
+    sense = getcursense(constr)
+    if nb_inf_sources == 0
+        if (sense != Greater) && alg_data.cur_max_slack[constr.id] < -0.0001
+            return true
+        elseif (sense == Greater) && alg_data.cur_max_slack[constr.id] <= -0.0001
+            #add_to_preprocessing_list(alg, constr)
```
[the remark this answers is not in the document]
Comment (vitornesello) [reply src/algorithms/preprocessing.jl]: This is normal.

**Comment by vitornesello** [comment]
I will merge because : 
- tests and coverage are OK
- Almost all of the comments are taken into account
- This code will be re-worked to add functionality to support benders and nested decomposition

**Pull request merged**